File: src/jvmTest/kotlin/org/snakeyaml/engine/usecases/env/EnvFormatTest.java
/*
 * Copyright (c) 2018, SnakeYAML
 *
 * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except
 * in compliance with the License. You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software distributed under the License
 * is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
 * or implied. See the License for the specific language governing permissions and limitations under
 * the License.
 */
package org.snakeyaml.engine.usecases.env;

import org.junit.jupiter.api.Test;
import it.krzeminski.snakeyaml.engine.kmp.resolver.BaseScalarResolver;

import java.util.regex.Matcher;
import java.util.regex.Pattern;

import static org.junit.jupiter.api.Assertions.*;

/**
 * {@code ${VARIABLE:-default}} evaluates to default if {@code VARIABLE} is unset or empty in the environment.
 * {@code ${VARIABLE-default}} evaluates to default only if VARIABLE is unset in the environment.
 * <p>
 * Similarly, the following syntax allows you to specify mandatory variables:
 * <p>
 * {@code ${VARIABLE:?err}} exits with an error message containing err if {@code VARIABLE} is unset or empty in the
 * environment. {@code ${VARIABLE?err}} exits with an error message containing err if {@code VARIABLE} is unset in
 * the environment.
 */
@org.junit.jupiter.api.Tag("fast")
public class EnvFormatTest {

  public static final Pattern ENV_FORMAT = BaseScalarResolver.ENV_FORMAT.toPattern();

  @Test
  public void testMatchBasic() {
    assertTrue(ENV_FORMAT.matcher("${V}").matches());
    assertTrue(ENV_FORMAT.matcher("${PATH}").matches());
    assertTrue(ENV_FORMAT.matcher("${VARIABLE}").matches());
    assertTrue(ENV_FORMAT.matcher("${ VARIABLE}").matches());
    assertTrue(ENV_FORMAT.matcher("${ VARIABLE}").matches());
    assertTrue(ENV_FORMAT.matcher("${ VARIABLE }").matches());
    assertTrue(ENV_FORMAT.matcher("${ VARIABLE}").matches());
    assertTrue(ENV_FORMAT.matcher("${\tVARIABLE  }").matches());

    Matcher matcher = ENV_FORMAT.matcher("${VARIABLE}");
    assertTrue(matcher.matches(), "expect that ENV_FORMAT matches '${VARIABLE}'");
    assertEquals("VARIABLE", matcher.group(1));
    assertNull(matcher.group(2));
    assertNull(matcher.group(3));

    assertFalse(ENV_FORMAT.matcher("${VARI ABLE}").matches());
  }

  @Test
  public void testMatchDefault() {
    assertTrue(ENV_FORMAT.matcher("${VARIABLE-default}").matches());
    assertTrue(ENV_FORMAT.matcher("${ VARIABLE-default}").matches());
    assertTrue(ENV_FORMAT.matcher("${ VARIABLE-default }").matches());
    assertTrue(ENV_FORMAT.matcher("${ VARIABLE-default}").matches());
    assertTrue(ENV_FORMAT.matcher("${ VARIABLE-}").matches());

    Matcher matcher = ENV_FORMAT.matcher("${VARIABLE-default}");
    assertTrue(matcher.matches(), "expect that ENV_FORMAT matches '${VARIABLE-default}'");
    assertEquals("VARIABLE", matcher.group(1));
    assertEquals("-", matcher.group(2));
    assertEquals("default", matcher.group(3));

    assertFalse(ENV_FORMAT.matcher("${VARIABLE -default}").matches());
    assertFalse(ENV_FORMAT.matcher("${VARIABLE - default}").matches());
    assertFalse(ENV_FORMAT.matcher("${VARIABLE -default}").matches());
  }

  @Test
  public void testMatchDefaultOrEmpty() {
    assertTrue(ENV_FORMAT.matcher("${VARIABLE:-default}").matches());
    assertTrue(ENV_FORMAT.matcher("${ VARIABLE:-default }").matches());
    assertTrue(ENV_FORMAT.matcher("${ VARIABLE:-}").matches());

    Matcher matcher = ENV_FORMAT.matcher("${VARIABLE:-default}");
    assertTrue(matcher.matches(), "expect that ENV_FORMAT matches '${VARIABLE:-default}'");
    assertEquals("VARIABLE", matcher.group(1));
    assertEquals(":-", matcher.group(2));
    assertEquals("default", matcher.group(3));

    assertFalse(ENV_FORMAT.matcher("${VARIABLE :-default}").matches());
    assertFalse(ENV_FORMAT.matcher("${VARIABLE : -default}").matches());
    assertFalse(ENV_FORMAT.matcher("${VARIABLE : - default}").matches());
  }

  @Test
  public void testMatchErrorDefaultOrEmpty() {
    assertTrue(ENV_FORMAT.matcher("${VARIABLE:?err}").matches());
    assertTrue(ENV_FORMAT.matcher("${ VARIABLE:?err }").matches());
    assertTrue(ENV_FORMAT.matcher("${ VARIABLE:? }").matches());

    Matcher matcher = ENV_FORMAT.matcher("${VARIABLE:?err}");
    assertTrue(matcher.matches(), "expect that ENV_FORMAT matches '${VARIABLE:?err}'");
    assertEquals("VARIABLE", matcher.group(1));
    assertEquals(":?", matcher.group(2));
    assertEquals("err", matcher.group(3));

    assertFalse(ENV_FORMAT.matcher("${ VARIABLE :?err }").matches());
    assertFalse(ENV_FORMAT.matcher("${ VARIABLE : ?err }").matches());
    assertFalse(ENV_FORMAT.matcher("${ VARIABLE : ? err }").matches());
  }

  @Test
  public void testMatchErrorDefault() {
    assertTrue(ENV_FORMAT.matcher("${VARIABLE?err}").matches());
    assertTrue(ENV_FORMAT.matcher("${ VARIABLE:?err }").matches());
    assertTrue(ENV_FORMAT.matcher("${ VARIABLE:?}").matches());

    Matcher matcher = ENV_FORMAT.matcher("${ VARIABLE?err }");
    assertTrue(matcher.matches(), "expect that ENV_FORMAT matches '${ VARIABLE?err }'");
    assertEquals("VARIABLE", matcher.group(1));
    assertEquals("?", matcher.group(2));
    assertEquals("err", matcher.group(3));

    assertFalse(ENV_FORMAT.matcher("${ VARIABLE ?err }").matches());
    assertFalse(ENV_FORMAT.matcher("${ VARIABLE ?err }").matches());
    assertFalse(ENV_FORMAT.matcher("${ VARIABLE ? err }").matches());
  }
}

File: src/commonTest/kotlin/it/krzeminski/snakeyaml/engine/kmp/usecases/env/EnvFormatTest.kt
package it.krzeminski.snakeyaml.engine.kmp.usecases.env

import io.kotest.core.spec.style.FunSpec
import io.kotest.matchers.nulls.shouldNotBeNull
import io.kotest.matchers.shouldBe
import it.krzeminski.snakeyaml.engine.kmp.resolver.BaseScalarResolver

/**
 * {@code ${VARIABLE:-default}} evaluates to default if {@code VARIABLE} is unset or empty in the environment.
 * {@code ${VARIABLE-default}} evaluates to default only if VARIABLE is unset in the environment.
 *
 * Similarly, the following syntax allows you to specify mandatory variables:
 *
 * {@code ${VARIABLE:?err}} exits with an error message containing err if {@code VARIABLE} is unset or empty in the
 * environment. {@code ${VARIABLE?err}} exits with an error message containing err if {@code VARIABLE} is unset in
 * the environment.
 */
class EnvFormatTest : FunSpec({
    test("match basic") {
        val envFormat = BaseScalarResolver.ENV_FORMAT

        envFormat.matches($$"${V}") shouldBe true
        envFormat.matches($$"${PATH}") shouldBe true
        envFormat.matches($$"${VARIABLE}") shouldBe true
        envFormat.matches($$"${ VARIABLE}") shouldBe true
        envFormat.matches($$"${ VARIABLE }") shouldBe true
        envFormat.matches($$"${\tVARIABLE  }") shouldBe true

        val matcher = envFormat.matchEntire($$"${VARIABLE}")
        matcher.shouldNotBeNull()
        matcher.groups[1]?.value shouldBe "VARIABLE"
        matcher.groups[2]?.value shouldBe null
        matcher.groups[3]?.value shouldBe null

        envFormat.matches($$"${VARI ABLE}") shouldBe false
    }

    test("match default") {
        val envFormat = BaseScalarResolver.ENV_FORMAT

        envFormat.matches($$"${VARIABLE-default}") shouldBe true
        envFormat.matches($$"${ VARIABLE-default}") shouldBe true
        envFormat.matches($$"${ VARIABLE-default }") shouldBe true
        envFormat.matches($$"${ VARIABLE-}") shouldBe true

        val matcher = envFormat.matchEntire($$"${VARIABLE-default}")
        matcher.shouldNotBeNull()
        matcher.groups[1]?.value shouldBe "VARIABLE"
        matcher.groups[2]?.value shouldBe "-"
        matcher.groups[3]?.value shouldBe "default"

        envFormat.matches($$"${VARIABLE -default}") shouldBe false
        envFormat.matches($$"${VARIABLE - default}") shouldBe false
    }

    test("match default or empty") {
        val envFormat = BaseScalarResolver.ENV_FORMAT

        envFormat.matches($$"${VARIABLE:-default}") shouldBe true
        envFormat.matches($$"${ VARIABLE:-default }") shouldBe true
        envFormat.matches($$"${ VARIABLE:-}") shouldBe true

        val matcher = envFormat.matchEntire($$"${VARIABLE:-default}")
        matcher.shouldNotBeNull()
        matcher.groups[1]?.value shouldBe "VARIABLE"
        matcher.groups[2]?.value shouldBe ":-"
        matcher.groups[3]?.value shouldBe "default"

        envFormat.matches($$"${VARIABLE :-default}") shouldBe false
        envFormat.matches($$"${VARIABLE : -default}") shouldBe false
        envFormat.matches($$"${VARIABLE : - default}") shouldBe false
    }

    test("match error default or empty") {
        val envFormat = BaseScalarResolver.ENV_FORMAT

        envFormat.matches($$"${VARIABLE:?err}") shouldBe true
        envFormat.matches($$"${ VARIABLE:?err }") shouldBe true
        envFormat.matches($$"${ VARIABLE:? }") shouldBe true

        val matcher = envFormat.matchEntire($$"${VARIABLE:?err}")
        matcher.shouldNotBeNull()
        matcher.groups[1]?.value shouldBe "VARIABLE"
        matcher.groups[2]?.value shouldBe ":?"
        matcher.groups[3]?.value shouldBe "err"

        envFormat.matches($$"${ VARIABLE :?err }") shouldBe false
        envFormat.matches($$"${ VARIABLE : ?err }") shouldBe false
        envFormat.matches($$"${ VARIABLE : ? err }") shouldBe false
    }

    test("match error default") {
        val envFormat = BaseScalarResolver.ENV_FORMAT

        envFormat.matches($$"${VARIABLE?err}") shouldBe true
        envFormat.matches($$"${ VARIABLE:?err }") shouldBe true
        envFormat.matches($$"${ VARIABLE:?}") shouldBe true

        val matcher = envFormat.matchEntire($$"${ VARIABLE?err }")
        matcher.shouldNotBeNull()
        matcher.groups[1]?.value shouldBe "VARIABLE"
        matcher.groups[2]?.value shouldBe "?"
        matcher.groups[3]?.value shouldBe "err"

        envFormat.matches($$"${ VARIABLE ?err }") shouldBe false
        envFormat.matches($$"${ VARIABLE ? err }") shouldBe false
    }
})
